Joystickinput settargets2

LorenzMeier opened the pull request:
@thomasgubler Can you give some recommendations how to best integrate this branch? Is this the better solution compared to joystickinput_settargets?

Susurrus commented:
Sounds like this branch needs some cleanup/more work before a PR. Should we close this until then?

thomasgubler commented:
I think this is still only branch which enables joystick offboard control with px4. Hence I would prefer to leave it here. Someone who needs joysytick control can clean it up eventually.

DonLakeFlyer commented:
Leave the branch but close the pull?

thomasgubler commented:
Isn't the open PR more obvious for someone who wants to pick up this work? As an open source project we need to accept that sometimes works is done at a slower pace (for whatever reason).

BTW we have 287 open issues and only 8 PR. If someone feels like doing house cleaning I would start with the issues. ;-)

LorenzMeier commented:
@Susurrus I have seen that some open source projects do not allow feature requests to be filed as issues. To me that makes some sense. I know that you're trying to help by giving a lot of good feedback via issues, but in the current workflow the older ones or the feature request ones just consume time to read and handle. It would be great if you could sort through them and kill all which are not direct bug reports or a short-term low hanging fruit you're planning to attack yourself or someone volunteers to fix.

DonLakeFlyer commented:
If it still needs work, then might as well leave it open. You get better tracking with a pull than just a separate branch and an issue. Could use more details as to what exactly it was and why it's here. It was a bit mysterious to anyone other than Lorenz and Thomas as to what exactly this was until a few comments ago.

thomasgubler commented:
See the comments here https://github.com/mavlink/qgroundcontrol/pull/813 which got removed again by revert on master.
This PR is basically the revert of the revert of #813

DonLakeFlyer commented:
Thanks Thomas.

thomasgubler commented:
* rebased
  * re-introduced manual control, is now a joystick mode like attitude or position.
  * quickly tested manual control mode with PX4 HIL.
  * some small fixes and style improvements

LorenzMeier commented:
I've rebased this again and squashed it into a single commit. I want to get this onto master as part of an effort to make Joysticks / Gamepads a 100% replacement for the remote control for SITL environments.

Susurrus commented on src/uas/UAS.cc:
```diff
@@ -2410,56 +2410,198 @@ void UAS::setManualControlCommands(float roll, float pitch, float yaw, float thr
+            const float fx = -dcm[0][2] * thrust;
+            const float fy = -dcm[1][2] * thrust;
+            const float fz = -dcm[2][2] * thrust;
+            uint16_t typeMask = (3<<10)|(7<<3)|(7<<0)|(1<<9); // select only FORCE control (disable everything else)
```
This is a super-confusing way to write out a typemask. It'd be better to just write it out as hex if all the parts don't just look like `1<<X`, especially since these bits aren't assigned in a logical order.

Susurrus commented on src/uas/UAS.cc:
```diff
@@ -2410,56 +2410,198 @@ void UAS::setManualControlCommands(float roll, float pitch, float yaw, float thr
+                }
+
+                // Now if we should trigger an update, let's do that
+                if (sendCommand) {
```
Shouldn't be this be done for all joystick modes?

LorenzMeier replied:
No, because the other modes send direct set points to the vehicle, which has an internal timeout. The joystick is acting like the high-level companion computer in these modes, and much higher updates rates are expected by the safety timeout logic in those cases.

Susurrus commented on src/uas/UAS.cc:
```diff
@@ -2410,56 +2410,198 @@ void UAS::setManualControlCommands(float roll, float pitch, float yaw, float thr
+            static float vy = 0;
+            static float vz = 0;
+            static float yawrate = 0;
+            //XXX: find decent scaling
```
Can determining these scalings only be done once this commit is merged or are they actually not necessary? Ideally this and the other XXX comments are resolved one way or another before its merged.

Susurrus commented:
Have manual control been tested with this PR? It's really important to my project, but I'm defending next week and my dissertation is due after that so I don't have time to test this PR on my own within the next couple of weeks.

LorenzMeier commented:
@Susurrus Testing now, but I don't recommend to use master for thesis defenses in general ;-).

LorenzMeier commented:
Sweet. Works as advertised!

LorenzMeier commented:
@Susurrus Thanks for the review! I went for the current type mask since I still found it easier to interpret than plain hex. I can transform small hex numbers in my head, but not a full mask.

thomasgubler commented:
:clap: thanks for getting this merged!
